`azurerm_windows_web_app`, `azurerm_windows_web_app_slot`, `azurerm_linux_web_app`, `azurerm_linux _web_app_slot` - fix auto heal slow request with path issue

xiaxyi opened the pull request:
In current provider, we only allow user to set the slow request rule with one path, but slow request rule with multiple paths is also supported by api, checking the property - [slowRequestsWithPath](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/web/resource-manager/Microsoft.Web/stable/2021-02-01/CommonDefinitions.json#L437):

The previous PR #18227 was closed but the issue was not fixed.

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/19256
fix https://github.com/hashicorp/terraform-provider-azurerm/issues/22492

Local Acc test:
```
--- PASS: TestAccLinuxWebAppSlot_withAutoHealSlowRequestWithPath (428.62s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    429.724s
--- PASS: TestAccWindowsWebAppSlot_withAutoHealSlowRequestWithPath (420.55s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    421.585s
--- PASS: TestAccLinuxWebApp_withAutoHealRulesSlowRequestWithPath (399.84s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    400.520s
--- PASS: TestAccWindowsWebApp_withAutoHealRulesSlowRequestWithPath (403.73s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    404.528s
```

![image](https://user-images.githubusercontent.com/92154856/212834517-34b9aba4-f50d-4aa5-a8d8-11435205958b.png)

pietersap commented:
I can confirm that the linked issue 17862 is not fixed yet, even though the previous PR was closed because it was supposedly superseded by another one.

Tried today in Azure provider 3.39.1, the autoheal settings (sub_status and win32_status) are not set, and behaviour is still the same as described in the original issue.

jackofallops reviewed (changes requested):
Hi @xiaxyi - Thanks for this and apologies again for the delayed review.

I think we have a design issue here. The two properties are present in the swagger, and as far as I can tell, both should be maintained / supported? Which means we should be able to simply add this as a new property? possibly as `ConflictsWith` on `slow_requests` if appropriate to only have one? WDYT?

xiaxyi commented:
Thanks @jackofallops for the suggestion, I didn't remove the `slow_request` property and I did added the `slow_request_with_path` as a new property. Not sure what do you mean by "add it as a new property"?

pietersap commented:
Looking forward to this merge

jackofallops commented on website/docs/r/windows_web_app.html.markdown:
```diff
@@ -512,6 +512,20 @@ A `slow_request` block supports the following:
 
 * `path` - (Optional) The path for which this slow request rule applies.
```
We should remove this from the docs as is it being deprecated
```suggestion
```

xiaxyi replied:
I added the deprecation message right below this property, shall we still remove it from the docs?

jackofallops reviewed (changes requested):
Hi @xiaxyi - Thanks for this. I've left some comments below if you can take a look?

Thanks!

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -44,6 +46,13 @@ type AutoHealSlowRequest struct {
 	Path      string `tfschema:"path"`
 }
 
+type AutoHealSlowRequestWithPath struct {
+	TimeTaken string `tfschema:"time_taken"`
+	Interval  string `tfschema:"interval"`
+	Count     int    `tfschema:"count"`
+	Path      string `tfschema:"path"`
+}
```
Thanks @jackofallops for the comment, I removed the path in 4.0 provider and the schema in 4.0 provider is defined as:
```
				"slow_request": {
					Type:     pluginsdk.TypeList,
					Computed: true,
					Elem: &pluginsdk.Resource{
						Schema: map[string]*pluginsdk.Schema{
							"time_taken": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"interval": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"count": {
								Type:     pluginsdk.TypeInt,
								Computed: true,
							},
						},
					},
				},

				"slow_request_with_path": {
					Type:     pluginsdk.TypeList,
					Computed: true,
					Elem: &pluginsdk.Resource{
						Schema: map[string]*pluginsdk.Schema{
							"time_taken": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"interval": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"count": {
								Type:     pluginsdk.TypeInt,
								Computed: true,
							},

							"path": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},
						},
					},
				},
			},
		},
	}

```

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -274,6 +283,32 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
+							"time_taken": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -274,6 +283,32 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
+							"interval": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
+					"time_taken": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
+					"interval": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/auto_heal.go:
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
+						ValidateFunc: validation.IntAtLeast(1),
+					},
+
+					"path": {
```
done

xiaxyi commented on internal/services/appservice/helpers/linux_web_app_schema.go:
```diff
@@ -615,6 +617,32 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
+							"time_taken": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/linux_web_app_schema.go:
```diff
@@ -615,6 +617,32 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
+							"interval": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/linux_web_app_schema.go:
```diff
@@ -627,10 +655,45 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
+					"time_taken": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/helpers/linux_web_app_schema.go:
```diff
@@ -627,10 +655,45 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
+					"interval": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
done

xiaxyi commented on internal/services/appservice/linux_web_app_resource_test.go:
```diff
@@ -2734,6 +2749,50 @@ resource "azurerm_linux_web_app" "test" {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
removed

xiaxyi commented on internal/services/appservice/linux_web_app_slot_resource_test.go:
```diff
@@ -1124,6 +1138,49 @@ resource "azurerm_linux_web_app_slot" "test" {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
removed

xiaxyi commented on internal/services/appservice/windows_web_app_resource_test.go:
```diff
@@ -2606,6 +2621,50 @@ resource "azurerm_windows_web_app" "test" {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
removed

xiaxyi commented on internal/services/appservice/windows_web_app_slot_resource_test.go:
```diff
@@ -964,6 +978,49 @@ resource "azurerm_windows_web_app_slot" "test" {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
removed

xiaxyi commented:
Thanks @jackofallops for the review, I updated the code, added the validation and deprecation message. 

Regarding the documentation, shall we still remove the `path` field even though I added the deprecation message right below the property?

jackofallops commented:
Hi @xiaxyi - Apologies for the delay in re-review. I have found there's a crash behind the feature flag here, which I don't believe can be addressed in this PR as the schema and the underlying go structs will no longer be compatible when the feature flag is set to true, so I'm going to mark this as `blocked` for now whilst the team investigates how to mitigate this.

xiaxyi commented:
@jackofallops I don't see any crash when running the tests, what about from your side? I see Tom removed the Block tag, would you like to share some updates with me? Thanks!

katbyte commented:
Now that the API has been updated and resources migrated can you please resolve the conflicts here?

jackofallops commented on website/docs/r/linux_web_app_slot.html.markdown:
```diff
@@ -790,7 +804,9 @@ A `trigger` block supports the following:
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

jackofallops commented on website/docs/r/windows_web_app.html.markdown:
```diff
@@ -773,6 +773,20 @@ A `slow_request` block supports the following:
 
 * `path` - (Optional) The path for which this slow request rule applies.
 
+~> **NOTE:** `path` in `slow_request` block will be deprecated in 4.0 provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```
```suggestion
~> **NOTE:** The `path` property in the `slow_request` block is deprecated and will be removed in 4.0 of provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```

jackofallops commented on website/docs/r/windows_web_app.html.markdown:
```diff
@@ -821,7 +835,9 @@ A `trigger` block supports the following:
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

jackofallops commented on website/docs/r/windows_web_app_slot.html.markdown:
```diff
@@ -770,6 +770,20 @@ A `slow_request` block supports the following:
 
 * `path` - (Optional) The path for which this slow request rule applies.
 
+~> **NOTE:** `path` in `slow_request` block will be deprecated in 4.0 provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```
```suggestion
~> **NOTE:** The `path` property in the `slow_request` block is deprecated and will be removed in 4.0 of provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```

jackofallops commented on website/docs/r/windows_web_app_slot.html.markdown:
```diff
@@ -810,7 +824,9 @@ A `trigger` block supports the following:
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

jackofallops reviewed (changes requested):
Hi @xiaxyi - Thanks for this PR, and sorry it fell between the cracks for so long. There's some changes to take a look at below, but if you can get those sorted, I think this will be good to go.

Thanks!

xiaxyi commented:
@jackofallops Thanks for the review, I updated the code as comments. Let me know if you have any other feedbacks.

jackofallops reviewed (approved):
Thanks @xiaxyi - Apologies for the _very_ long tail on this one. LGTM now 👍

jackofallops commented:
Tests looking fine, failures are transient errors:

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/11830746/96968a8d-92aa-494a-8708-abbc03b9e97e)
